### ISSUE TITLE
get_viewport().warp_mouse does not works on linux

### ISSUE THREAD

**Issue by gdesecrate** [godotengine/godot#20410]
**Godot version:**
2.1

**OS/device including version:**
Linux/Windows

**Issue description:**
On linux mouse behavior differs from windows.
Seems like no warping is happening while on windows warp is automatical after calling Input.set_mouse_mode(Input.MOUSE_MODE_CAPTURED); 

**Steps to reproduce:**
#on linux
print(get_viewport().get_mouse_pos())
get_viewport().warp_mouse(get_viewport().get_rect().size/2.0)
print(get_viewport().get_mouse_pos())

**Comment by akien-mga** [comment]
On Linux those things are often updated in the next frame (at least in the way the X11 queries are done currently), so you might need to wait one frame to get the new value. Try this:
```
print(get_viewport().get_mouse_pos())
get_viewport().warp_mouse(get_viewport().get_rect().size/2.0)
yield(get_tree(), "idle_frame")
print(get_viewport().get_mouse_pos())
```

**Comment by akien-mga** [comment]
> Godot version: 2.1

Is it 2.1 (from 2016) or a self-compiled build from the 2.1 branch? (~2.1.5)

Can you reproduce the issue on Godot 3.0.x too?

**Comment by gdesecrate** [comment]
Godot_v2.1.5-rc2_x11
I do not have 3.x and do not plan to install it until it is stable and got gles2.

The issue makes game loze zero-point. So when you start moving mouse the last motion frame from previous motions affect 1st frame on new motion.

Really ruins FPS/TPS experience even if you filter it hard.

**Comment by cosmicchipsocket** [comment]
Not sure if the mouse issues are the same as I have been experiencing, but I have an open PR to use raw mouse motion in the 3.1 branch at #19191.

**Comment by aaronfranke** [comment]
@gdesecrate I understand that you have reasons for sticking with Godot 2.1 for now, but it would be easier to fix the problem if we knew more information, such as if it's a problem in 3.x or not.

**Comment by gdesecrate** [comment]
@aaronfranke I don't understand why you do make drama instead of launching the code I provided. Since you prob too lazy to scroll up - i'll repost. 3.x have different procedure header's format - I removed them for your convinience. 

#in fixed process func
print(get_viewport().get_mouse_pos())
get_viewport().warp_mouse(get_viewport().get_rect().size/2.0)
print(get_viewport().get_mouse_pos())
print("^If numbers up do match WHEN you DO move mouse then warp is broken.")
#...
#at spawn
Input.set_mouse_mode(Input.MOUSE_MODE_CAPTURED)

**Comment by akien-mga** [comment]
@gdesecrate Please stay respectful of other users.

**Comment by gdesecrate** [comment]
@akien-mga He sounded like in a panic. And I reposted issue for code for him to test. Why those calls platform-depended in a first place?

**Comment by akien-mga** [comment]
> He sounded like in a panic.

I don't think so, but even if it were the case, I don't think being disrespectful to people being "like in a panic" will help them feel better/less panicked.

>  And I reposted issue for code for him to test. Why those calls platform-depended in a first place?

Anything related to input and window is platform dependent by nature.

> If there was something like a
get_viewport().get_mouse_speed()
for fixed process I even woudn't noticed that issue

How about `Input.get_mouse_speed()`? http://docs.godotengine.org/en/2.1/classes/class_input.html?highlight=mouse_speed#class-input-get-mouse-speed

**Comment by gdesecrate** [comment]
@akien-mga Your `Input.get_mouse_speed()` is uselles unless I additionaly track if mouse moving.

I just want every frame in my _fixed_process() func to get Vector2 how much mousemoved relatively to previous frame. And because "warp" so broken I get like data from old mouse movements. 
Like last frame of movement gets to new mouse movement even if a noticeable time was spent. It really bad for FPS controls.

All was fine till I noticed that setting mouse capture does warps mouse on windows and does not on linux. 

> Anything related to input and window is platform dependent by nature.
Never had that issue with SDL. It seems like you are trying to impose "your nature" of some sort there.

**Comment by akien-mga** [comment]
>> Anything related to input and window is platform dependent by nature.
>
> Never had that issue with SDL. It seems like you are trying to impose "your nature" of some sort there.

The fact that it works reliably on all platforms on SDL doesn't contradict the fact that *input and window management is platform-dependent by nature*.

And again, I would like you to stop accusing/disrespecting people who are trying to help. If you can't, I'll just lock the issue and you can try to figure things out on your own (or use SDL).

**Comment by akien-mga** [comment]
> I just need a Vector2 of mouse movement for every tick. You are the one that brings up "every platform is different"-thing, while the Input layer must be transparent for engine's end user.

It's not "me that brings up", I just answered your question about why those calls are platform-dependent. They are, by nature, that's all. It's hard science, it's not an opinion. It doesn't solve the issue, and doesn't aim to, it was just an answer to your question.

> And when you trying to tell me "use that function that does not tracks if mouse stopped" is that some sort of you disrespecting me?

No, it was a suggestion. I did not mean that it would work, I suggested to try it.

**Comment by gdesecrate** [comment]
> No, it was a suggestion. I did not mean that it would work, I suggested to try it.
I already did multiple times in the past. Bad coordinate feed, requires filtering and flagging the input events.

**Comment by gdesecrate** [comment]
By the way if I remove the warp - it works fine. But warp is added for window's build by default.

**Comment by guilhermefelipecgs** [comment]
Yep, this was already fixed for Linux.

**Comment by akien-mga** [comment]
Thanks, closing as fixed in 3.1 then.

**Comment by Skaruts** [comment]
Has this got broken in the meanwhile? I'm getting this issue on Windows with Godot 3.1.1, as well as the nighly build I downloaded just now (77e894768).
